shared_testutil/mockdagstorewrapper.go

```diff
@@ -0,0 +1,136 @@
+package shared_testutil
+
+import (
+	"context"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/dagstore"
+
+	"github.com/filecoin-project/go-fil-markets/piecestore"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/stores"
+)
+
+type registration struct {
+	CarPath   string
+	EagerInit bool
+}
+
+// MockDagStoreWrapper is used to mock out the DAG store wrapper operations
+// for the tests.
+// It simulates getting deal info from a piece store and unsealing the data for
+// the deal from a retrieval provider node.
+type MockDagStoreWrapper struct {
+	pieceStore piecestore.PieceStore
+	rpn        retrievalmarket.RetrievalProviderNode
+
+	lk            sync.Mutex
+	registrations map[cid.Cid]registration
+}
+
+var _ stores.DAGStoreWrapper = (*MockDagStoreWrapper)(nil)
+
+func NewMockDagStoreWrapper(pieceStore piecestore.PieceStore, rpn retrievalmarket.RetrievalProviderNode) *MockDagStoreWrapper {
+	return &MockDagStoreWrapper{
+		pieceStore:    pieceStore,
+		rpn:           rpn,
+		registrations: make(map[cid.Cid]registration),
+	}
+}
+
+func (m *MockDagStoreWrapper) RegisterShard(ctx context.Context, pieceCid cid.Cid, carPath string, eagerInit bool, resch chan dagstore.ShardResult) error {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	m.registrations[pieceCid] = registration{
+		CarPath:   carPath,
+		EagerInit: eagerInit,
+	}
+
+	resch <- dagstore.ShardResult{}
+	return nil
+}
+
+func (m *MockDagStoreWrapper) MigrateDeals(ctx context.Context, deals []storagemarket.MinerDeal) (bool, error) {
+	return true, nil
+}
+
+func (m *MockDagStoreWrapper) LenRegistrations() int {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	return len(m.registrations)
+}
+
+func (m *MockDagStoreWrapper) GetRegistration(pieceCid cid.Cid) (registration, bool) {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	reg, ok := m.registrations[pieceCid]
+	return reg, ok
+}
+
+func (m *MockDagStoreWrapper) ClearRegistrations() {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	m.registrations = make(map[cid.Cid]registration)
+}
+
+func (m *MockDagStoreWrapper) LoadShard(ctx context.Context, pieceCid cid.Cid) (stores.ClosableBlockstore, error) {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	_, ok := m.registrations[pieceCid]
+	if !ok {
+		return nil, xerrors.Errorf("no shard for piece CID %s", pieceCid)
+	}
+
+	// Get the piece info from the piece store
+	pi, err := m.pieceStore.GetPieceInfo(pieceCid)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unseal the sector data for the deal
+	deal := pi.Deals[0]
+	r, err := m.rpn.UnsealSector(ctx, deal.SectorID, deal.Offset.Unpadded(), deal.Length.Unpadded())
```

Comment:
Two things:
1. Why go through all this chicanery for something that's a mock implementation to begin with?
2. My guess from reading the code is that the dagstore hides the Unsealing behind LoadShard. (or some wrapper in Lotus does) and as far as I can tell UnsealSector, the method on the retrieval provider node, is no longer used, except oddly in this test fixture. If this is the case, I would say, remove it. We should not have methods on the node that aren't used, as it means a node implementer must support them.

Reply:
You're right that the dag store now takes care of unsealing internally. So in order to test that the deal flow works end to end, including the unsealing part, we implemented this mock dag store wrapper.

Reply:
I would say: a mock implementation is a mock implementation, and having it do fake implementations that themselves call mock functions (UnsealSector is a mock implementation) doesn't actually add value to testing code.

The assumption is that the boundary of testing here is what's in the go-fil-markets library, that the dagstore is unit tested, and the rest is handled by lotus integration testing.

Reply:
However, I think the reason why this is needed is because even "inside" the test framework, the component that custodies the pieces is the `PieceStore`, and there's a mock for that. Same for the `SectorAccessor`. And a test that uses this test framework would reasonably assume that it can reuse those bits to mock out the piece and sector layers, and have those feed into the rest of the test framework. Unfortunately that leads to a MockDagStoreWrapper that actually holds a little bit of logic (which is not illegal IMO, because it's just satisfying the contract of the interface in the advertised manner), but I can't think of another way.